appinventor/appengine/src/com/google/appinventor/client/editor/simple/components/MockLineChartModel.java

```diff
@@ -0,0 +1,98 @@
+
+public class MockLineChartModel extends MockChartModel<ScatterDataset> {
+
+    public MockLineChartModel(Data chartData) {
+        super(chartData);
+
+        // Create the Data Series object
+        dataSeries = new ScatterDataset();
+
+        dataSeries.setFill(false);
+        dataSeries.setBorderWidth(1);
+        dataSeries.setLineTension(0);
+        dataSeries.setShowLine(true);
+
+        // Adds the Data Series to the Chart.
+        addDataSeriesToChart();
+    }
+
+    @Override
+    public void changeColor(String color) {
+        color = getHexColor(color);
+        dataSeries.setBackgroundColor(color);
+        dataSeries.setPointBackgroundColor(color);
+        dataSeries.setBorderColor(color);
+    }
+
+    @Override
+    public void setElements(String elements) {
+        String[] entries = elements.split(",");
+
+        // Create new list of Data Points
+        ArrayList<DataPoint> dataPoints = new ArrayList<DataPoint>();
+
+        // Since entries come in pairs, we add 2 on each iteration.
+        // Beginning from i = 1 instead of 0 to prevent out of bounds
+        // accesses.
+        for (int i = 1; i < entries.length; i += 2) {
+            try {
+                DataPoint dataPoint = new DataPoint();
+                dataPoint.setX(Double.parseDouble(entries[i-1]));
+                dataPoint.setY(Double.parseDouble(entries[i]));
+                dataPoints.add(dataPoint);
+            } catch (NumberFormatException e) {
+                return; // Wrong input. Do not update entries.
+            }
+        }
+
+        // No data points generated, fallback to default option.
+        if (dataPoints.isEmpty()) {
+            setDefaultElements(dataPoints);
+        } else {
+            // Since we are dealing with a Scatter Data Series, sorting
+            // is a must, because otherwise, the Chart will not look representative.
+            // Consider adding: (1, 2), (5, 3), (2, 5). We want the x = 2
+            // value to be continuous on the Line Chart, rather than
+            // going outside the Chart.
+            dataPoints.sort(Comparator.comparingDouble(DataPoint::getX));
+        }
+
+        // Set the data points to the actual Data Series
+        dataSeries.setDataPoints(dataPoints);
+    }
+
+    @Override
+    protected void setDefaultElements(List<DataPoint> dataPoints) {
+        final int points = 4; // Number of points to add
+
+        // TBD: Might change this in the future.
+        // Generally, this should not cause performance issues because typically there are not
+        // that many data points.
+        Optional<DataPoint> maxYPoint = chartData.getDatasets() // Get all the data sets
+                .stream() // Create a stream
+                .flatMap(l -> ((ScatterDataset)l).getDataPoints().stream()) // Flatten the nested lists to a List of data points
+                .max(Comparator.comparing(DataPoint::getY)); // Get the maximum data point value
```

Comment:
Currently, this is the approach I decided to go with to solve one issue in particular; When Data components are deleted, and added again, they should be added to the Chart such that Lines do not overlay each other. To do this, a new Line is generated such that it is on top of the maximum y value of the data.

Now, this causes a somewhat complex solution. Questions:
1.Is it okay to have this O(n * m) average complexity solution? It should not matter for small inputs, but this still has high potential to slow down machines.


2.Is using Java 8 functionality appropriate here? I used Streams because it made for an easy to understand and fast (in terms of time taken) solution.


3.Line 89; See reasoning on the implementation. Is this a bit of a 'magic number' implementation, or does it make sense?